auth/auth.go

```diff
@@ -0,0 +1,172 @@
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+// IAuth defines an interface for auth related functionality.
+type IAuth interface {
+	ReadAndSetKeys()
+	UserIsLoggedIn() gin.HandlerFunc
+	UserIsInRole(d database.IDatabase, roles []constants.Role) gin.HandlerFunc
+	CreateToken(user model.User) (string, error)
+}
+
+type Auth struct {
+}
+
+var (
+	signKey   *rsa.PrivateKey
+	verifyKey *rsa.PublicKey
+)
+
+// ReadAndSetKeys will read public and private RSA keys and create a key for signing JWTs.
+func (a *Auth) ReadAndSetKeys() {
+	a.setUpViper()
+
+	privKeyPath := viper.GetString("PRIVATE_KEY_PATH")
+	pubKeyPath := viper.GetString("PUBLIC_KEY_PATH")
+
+	signBytes, err := ioutil.ReadFile(privKeyPath)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"privateKeyPath": privKeyPath,
+		}).Fatal(err)
+	}
+
+	signKey, err = jwt.ParseRSAPrivateKeyFromPEM(signBytes)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	verifyBytes, err := ioutil.ReadFile(pubKeyPath)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"publiceKeyPath": pubKeyPath,
+		}).Fatal(err)
+	}
+
+	verifyKey, err = jwt.ParseRSAPublicKeyFromPEM(verifyBytes)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// UserIsLoggedIn will read the JWT in the request header and verify that it is legitimate.
+func (a *Auth) UserIsLoggedIn() gin.HandlerFunc {
+	return func(c *gin.Context) {
+
+		token, err := a.getToken(c)
+
+		if err != nil {
+			c.Abort()
+			return
+		}
+
+		if token.Valid {
+			// save token in context for use in other middleware
+			c.Set("token", token)
+		} else {
+			origin := c.GetHeader("Origin")
+			if ve, ok := err.(*jwt.ValidationError); ok {
+				if ve.Errors&jwt.ValidationErrorMalformed != 0 {
+					c.JSON(http.StatusForbidden, gin.H{"err": "token is malformed"})
+					log.Error(fmt.Sprintf("Attempt to access resources with malformed token from origin: %s", origin))
+				} else if ve.Errors&(jwt.ValidationErrorExpired) != 0 {
+					c.JSON(http.StatusForbidden, gin.H{"err": "token is expired"})
+				} else {
+					c.JSON(http.StatusForbidden, gin.H{"err": "error reading token"})
+					log.Error(fmt.Sprintf("Error accessing resources with token from origin %s", origin))
+				}
+			} else {
+				c.JSON(http.StatusForbidden, gin.H{"err": "error reading token"})
+				log.Error(fmt.Sprintf("Error accessing resources with token from origin %s", origin))
+			}
+			c.Abort()
+		}
+	}
+}
+
+// UserIsInRole accepts a list of roles and determines whether a user's role in a JWT is in that list.
+func (a *Auth) UserIsInRole(d database.IDatabase, roles []constants.Role) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var token interface{}
+		var ok bool
+
+		if token, ok = c.Get("token"); !ok {
+			c.JSON(http.StatusForbidden, gin.H{"err": "Error accessing token"})
+			c.Abort()
+			return
+		}
+
+		var userRole constants.Role
+		if claims, ok := token.(*jwt.Token).Claims.(jwt.MapClaims); ok {
+			userRole = constants.Role(claims["role"].(float64))
```

Comment:
float64? Maybe I'm remembering this wrong but I thought that was an int

Reply:
This one was weird but when trying to case to an int it game an error like `interface conversion: interface {} is float64, not int`